src/operator/delayWhen.ts

```diff
@@ -0,0 +1,155 @@
+import {Operator} from '../Operator';
+import {Subscriber} from '../Subscriber';
+import {Observable} from '../Observable';
+import {Subscription} from '../Subscription';
+
+import {tryCatch} from '../util/tryCatch';
+import {errorObject} from '../util/errorObject';
+
+import {OuterSubscriber} from '../OuterSubscriber';
+import {InnerSubscriber} from '../InnerSubscriber';
+import {subscribeToResult} from '../util/subscribeToResult';
+
+/**
+ * Returns an Observable that delays the emission of items from the source Observable
+ * by a subscription delay and a delay selector function for each element.
+ * @param {Function} selector function to retrieve a sequence indicating the delay for each given element.
+ * @param {Observable} sequence indicating the delay for the subscription to the source.
+ * @returns {Observable} an Observable that delays the emissions of the source Observable by the specified timeout or Date.
+ */
+
+export function delayWhen<T>(delayDurationSelector: (value: T) => Observable<any>,
+                             subscriptionDelay?: Observable<any>): Observable<T> {
+  if (subscriptionDelay) {
+    return new SubscriptionDelayObservable(this, subscriptionDelay)
+            .lift(new DelayWhenOperator(delayDurationSelector));
+  }
+  return this.lift(new DelayWhenOperator(delayDurationSelector));
+}
+
+class DelayWhenOperator<T> implements Operator<T, T> {
+  constructor(private delayDurationSelector: (value: T) => Observable<any>) {
+  }
+
+  call(subscriber: Subscriber<T>): Subscriber<T> {
+    return new DelayWhenSubscriber(subscriber, this.delayDurationSelector);
+  }
+}
+
+interface DelayNotiferSubContext<T> {
+  index: number;
+  subscription: Subscription;
+  value: T;
+}
+
+class DelayWhenSubscriber<T, R> extends OuterSubscriber<T, R> {
+  private completed: boolean = false;
+  private delayNotifiers: Array<DelayNotiferSubContext<T>> = [];
```

Comment:
now tried to use simple array instead of map, subscription deletion via lookup & splice.